File: Convert Sorted Array to Binary Search Tree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode *sortArrayToBST(vector<int> &num, int start, int end) {
        if (start > end) return NULL;
        
        int mid = start + (end - start) / 2;
        TreeNode *root = new TreeNode(num[mid]);
        
        TreeNode *leftchild = sortArrayToBST(num, start, mid - 1);
        TreeNode *rightchild = sortArrayToBST(num, mid + 1, end);
        
        root->left = leftchild;
        root->right = rightchild;
        
        return root;
    }
    TreeNode *sortedArrayToBST(vector<int> &num) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        return sortArrayToBST(num, 0, num.size() - 1);
    }
};

File: Convert Sorted Array to Binary Search Tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode SortedArrayToBST(int[] nums) {
        return SortedArrayToBST(nums, 0, nums.Length - 1);
    }
    
    private TreeNode SortedArrayToBST(int[] nums, int start, int end) {
        if (start > end) {
            return null;
        }
        
        var mid = start + (end - start) / 2;
        var root = new TreeNode(nums[mid]);
        root.left = SortedArrayToBST(nums, start, mid - 1);
        root.right = SortedArrayToBST(nums, mid + 1, end);
        
        return root;
    }
}
